### PR TITLE
Support lexicographic range queries (>, <, >=, <=) on TAG and TEXT indices

### PR DESCRIPTION
**Pull request by ShubhamTaple**
## Describe the changes in the pull request
Addresses the needs of this particular request: https://github.com/redis/redis/issues/14996

A clear and concise description of what the PR is solving, including:
1. Current: The Redis Query Engine supports comparison operators (`>, >=, <, <=, ==, !=`) only on `NUMERIC` fields. `TAG` and `TEXT` fields can only be queried via exact match `(@tag:{value})`, prefix/suffix patterns, or lexicographic ranges using the bracket syntax `([begin (end)`. There is no way to express lexicographic comparisons using the familiar operator syntax, which blocks patterns like keyset-based pagination on string-indexed fields.

2. Change: Extended the query parser (lexer + grammar) to accept string `TERM` values as operands to comparison operators. When the field is `TAG` or `TEXT`, the parser now creates `QN_LEXRANGE` nodes (which already had full evaluation support internally via `Query_EvalTagLexRangeNode` and `Query_EvalLexRangeNode`). Numeric fields continue to work exactly as before — the lexer uses `fast_float_strtod` to distinguish numbers from strings at tokenization time, and the parser dispatches to separate grammar rules (`param_num` vs `TERM`) with zero conflicts.

3. Outcome: Users can now write queries like `@name > Marsh`, `@city <= bangalore`, `@username >= shubham` `@username < aish` on `TAG` and `TEXT` indices. This directly enables keyset-based pagination and other lexicographic filtering patterns that were previously impossible without workarounds.

#### Which additional issues this PR fixes
None

#### Main objects this PR modified
1.`src/query_parser/v2/lexer.rl` — Extended ragel patterns for comparison operators to accept term operands; modified `RSQuery_ParseNumericOp_v2()` to emit `TERM` tokens when the value is not a valid number.

2. `src/query_parser/v2/parser.y` — Added 8 new grammar rules (`GT/GE/LT/LE/EQUALS/NOT_EQUAL × TERM`) that create `QN_LEXRANGE` nodes for `TAG` and `TEXT` fields.

3.`src/query.c` — Added `NewLexRangeNode()` helper to construct `QN_LEXRANGE AST` nodes from length-delimited strings.

4. `src/query_internal.h` — Declared `NewLexRangeNode()` so parser-generated code can call it.

5. `tests/pytests/test_lexrange.py` — Added 11 integration tests covering all operators on `TAG/TEXT` fields, edge cases, regression for numeric operators, keyset pagination, and combined filters.

#### Mark if applicable

- [x] This PR introduces API changes
- [ ] This PR introduces serialization changes

#### Release Notes

- [x] This PR requires release notes
- [ ] This PR does not require release notes

Support lexicographic range queries (`>, >=, <, <=, ==, !=`) on `TAG` and `TEXT` indices using the same operator syntax already available for `NUMERIC` fields. This enables patterns like keyset-based pagination on string-valued fields (e.g., `FT.SEARCH idx "@username > shubham" SORTBY username ASC LIMIT 0 10`).

---

> [!NOTE]
> **Medium Risk**
> Extends the v2 lexer/parser to reinterpret comparison operators with non-numeric RHS as lexicographic range filters for `TAG`/`TEXT`, which could subtly change how some previously-accepted mixed tokens were tokenized (e.g. leading-digit strings) and impact query behavior.
> 
> **Overview**
> Adds support for lexicographic comparisons on `TAG` and `TEXT` fields using standard operators (`>`, `>=`, `<`, `<=`, plus `==`/`!=` for term equality) by emitting `QN_LEXRANGE` nodes from the v2 parser.
> 
> The lexer now accepts `TERM` operands for modifier comparisons and tightens numeric tokenization to require full RHS consumption (so inputs like `5stars` become `TERM`, not a partially-parsed `NUMBER`). A new `NewLexRangeNode()` constructor is introduced, and a new Python test suite exercises the new operator syntax, regression coverage for numeric operators, and keyset-pagination style queries.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 02d32d8ec5cb273112d7e5ea753f2067bcd8c402. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/query_parser/v2/lexer.rl [hunk 1]
```diff
@@ -84,9 +84,15 @@ int RSQuery_ParseNumericOp_v2(void* pParser, int OperatorType, QueryToken tok,
...
+      char *ne = (char*)tok.s;
       tok.numval = fast_float_strtod(tok.s, &ne);
-      RSQuery_Parse_v2(pParser, NUMBER, tok, q);
+      if (ne > tok.s) {
+        RSQuery_Parse_v2(pParser, NUMBER, tok, q);
+      } else {
+        tok.numval = 0;
+        tok.len = te - tok.s;
+        RSQuery_Parse_v2(pParser, TERM, tok, q);
+      }
```
[the remark this answers is not in the document]
Comment (ShubhamTaple) [reply src/query_parser/v2/lexer.rl:101]: Addressed

**Comment by aradalvand** [comment]
Thank you @ShubhamTaple for this effort, appreciated.

@adrianoamaral @kei-nan Can you please take a look and let us know if there's a chance this could get merged?

**Comment by aradalvand** [comment]
@oshadmi Can you please take a look?

**Comment by aradalvand** [comment]
I'm so tired of pinging you @kei-nan

**Comment by aradalvand** [comment]
For fuck's sake, man.
The reason this "hasn't had activity" is because it's waiting for a review. Once again, @kei-nan, I'm pinging you.

File: src/query_parser/v2/lexer.rl [hunk 2]
```diff
@@ -84,9 +84,21 @@ int RSQuery_ParseNumericOp_v2(void* pParser, int OperatorType, QueryToken tok,
       }
       RSQuery_Parse_v2(pParser, ATTRIBUTE, tok, q);
     } else {
-      char *ne = (char*)te;
+      char *ne = (char*)tok.s;
```
Comment (nafraf) [inline src/query_parser/v2/lexer.rl:87]: Please regenerate the lexer/parser files and added them to the PR.
- `make parsers` 
- commit the regenerated lexer.c / parser.c

File: src/query.c [hunk 3]
```diff
@@ -337,6 +337,20 @@ QueryNode *NewMissingNode(const FieldSpec *field) {
   return ret;
 }
 
+QueryNode *NewLexRangeNode(const char *begin, size_t beginLen, bool includeBegin,
```
Comment (nafraf) [inline src/query.c:340]: We need to set `opts.fieldIndex` on TAG lex-range nodes.
This test reproduce the current error:
```sh
127.0.0.1:6379> DEBUG SET-ACTIVE-EXPIRE 0
OK
127.0.0.1:6379> FT.CREATE idx ON HASH SCHEMA first TAG second TAG
OK
127.0.0.1:6379> HSET doc:1 first aaa second bbb
(integer) 2
127.0.0.1:6379> HPEXPIRE doc:1 1 FIELDS 1 first
1) (integer) 1
127.0.0.1:6379> FT.SEARCH idx '@second:{bbb}' nocontent DIALECT 2
1) (integer) 1
2) "doc:1"
127.0.0.1:6379> FT.SEARCH idx '@second > aaa' nocontent DIALECT 2
1) (integer) 0 
```
The last query should return `doc:1`, but it's returning 0 results, it's querying on the first field.

File: tests/pytests/test_lexrange.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    # @name >= charlie -> charlie, dave, eve
+    res = env.cmd('ft.search', 'idx', '@name >= charlie', 'nocontent')
+    env.assertEqual(res[0], 3)
+    results = py2sorted(res[1:])
+    env.assertContains('doc3', results)
+    env.assertContains('doc4', results)
+    env.assertContains('doc5', results)
+
+
+def testTagLexRangeLT(env):
```
Comment (nafraf) [inline tests/pytests/test_lexrange.py:67]: Consider adding tests for `==` and `!=` operators.

File: src/query_parser/v2/lexer.rl [hunk 5]
```diff
@@ -84,9 +84,21 @@ int RSQuery_ParseNumericOp_v2(void* pParser, int OperatorType, QueryToken tok,
...
       tok.numval = fast_float_strtod(tok.s, &ne);
-      RSQuery_Parse_v2(pParser, NUMBER, tok, q);
+      /* Require full consumption of the RHS as a number: partial parses like
```
Comment (nafraf) [inline src/query_parser/v2/lexer.rl:89]: This is a breaking change, a working query becomes an error.
We'll need to decide if we include the fix only in the next major version.
Before the change:
```sh
> FT.EXPLAIN idx "@price>5stars"
INTERSECT {
  NUMERIC {5.000000 < @price <= inf}
  UNION { stars +star(expanded) star(expanded) }
}
```
New behavior:
```sh
> FT.SEARCH idx "@price>5stars" NOCONTENT
(error) SEARCH_SYNTAX Expected a NUMERIC or TAG or TEXT field at offset 0 near price
```
Additionally, consider fixing the error message:
`(error) SEARCH_SYNTAX Expected a NUMERIC`

File: src/query_parser/v2/lexer.rl [hunk 6]
```diff
@@ -125,12 +137,12 @@ term = (((any - (punct | cntrl | space | escape)) | escaped_character) | '_')+
...
 attr = '$'.term $ 1;
-mod_not_equal = '@'.term.(space*).'!='.(space*).(number|inf|size|('+'|'-')?.attr) $ 1;
-mod_equal = '@'.term.(space*).'=='.(space*).(number|inf|size|('+'|'-')?.attr) $ 1;
-mod_gt = '@'.term.(space*).'>'.(space*).(number|inf|size|('+'|'-')?.attr) $ 1;
-mod_ge = '@'.term.(space*).'>='.(space*).(number|inf|size|('+'|'-')?.attr) $ 1;
-mod_lt = '@'.term.(space*).'<'.(space*).(number|inf|size|('+'|'-')?.attr) $ 1;
-mod_le = '@'.term.(space*).'<='.(space*).(number|inf|size|('+'|'-')?.attr) $ 1;
+mod_not_equal = '@'.term.(space*).'!='.(space*).(number|inf|size|('+'|'-')?.attr|term) $ 1;
+mod_equal = '@'.term.(space*).'=='.(space*).(number|inf|size|('+'|'-')?.attr|term) $ 1;
+mod_gt = '@'.term.(space*).'>'.(space*).(number|inf|size|('+'|'-')?.attr|term) $ 1;
```
Comment (nafraf) [inline src/query_parser/v2/lexer.rl:142]: The lexer is classifying `nan` as number even when the fieldtype is TAG.
```sh
FT.CREATE idx on hash schema name tag
HSET doc1 name nan
HSET doc1 name nap

127.0.0.1:6380> FT.SEARCH idx '@name>=nap' DIALECT 2
1) (integer) 1
2) "doc1"
3) 1) "name"
   2) "nap"
127.0.0.1:6380> FT.SEARCH idx '@name>=nan' DIALECT 2
(error) SEARCH_SYNTAX Expected a NUMERIC field at offset 0 near name
```

**Review by nafraf — COMMENTED** [review]
I left some comments.
Please merge master branch to take the last changes.

**Comment by kei-nan** [comment]
I’m a bit worried about making `@name > foo` the public syntax here. It reads nicely, but it also puts string ranges on the same lexer path as numeric comparisons, and that seems to be where the awkward cases like `5stars` and `nan` show up.

What do you think about keeping the operator inside the normal field-filter shape instead?

```text
@name:>(foo)
@name:>=(foo)
@name:<(foo)
@name:<=(foo)

@name:>(bob) @name:<=(dave)
@city:>{"New York"}
@name:>($cursor)
```

To me this has a few advantages: the operator is clearly part of the `@field:` clause, it feels closer to the existing text/tag/range syntaxes, and it gives us a clean boundary for parsing the RHS as a string instead of trying to infer whether the operand is numeric or textual.

I’d probably document the parenthesized form as the main one. Maybe `@name:>foo` can be accepted as shorthand for simple values, but I’d avoid the space-separated `@name > foo` form.

I’d also consider leaving `==` and `!=` out of this PR unless parity with numeric operators is a hard requirement. TAG/TEXT already have exact-match and negated exact-match syntax, while the real missing piece for keyset pagination is `<`, `<=`, `>`, and `>=`.

**Comment by aradalvand** [comment]
@ShubhamTaple Are you available to update this PR based on the feedback?
If not, I could take over.

**Comment by oshadmi** [comment]
I suggest to guard this capability behind `ENABLE_UNSTABLE_FEATURES`
(see docs/CONTRIBUTING-unstable-features.md)

Need to define the full semantics, e.g., `@city:>(New York)` should fail with multiple TEXT terms?

**Comment by ShubhamTaple** [comment]
> @ShubhamTaple Are you available to update this PR based on the feedback? If not, I could take over.

@aradalvand I would be able to contribute to this PR  only on weekends (so it might get stretched across multiple weeks). If you are fine with it, I will continue to contribute to this PR.

If that doesn't work for you then feel free to take over.

**Comment by aradalvand** [comment]
I just created #11273 as a superset of this PR (because frankly I am desperate for this feature and @ShubhamTaple seemed busy this weekend).
@oshadmi @nafraf I addressed you guys's concerns. Please check out that PR.
